### PR TITLE
Add support for NWM 3.0 baseline output variables

### PR DESCRIPTION
**Pull request by GreyEvenson-NOAA**
# Purpose

The purpose of this PR is to expose Noah-OM variables via BMI to replicate the functionality of NWM 3.0. 

# Additions

The following model variables were added to `output_items` in `bmi/bmi_noahowp.f90`:

- `ACSNOM` -- The accumulated meltwater from bottom snow layer (mm) (NWM 3.0 output variable)
- `SNOWT_AVG`  -- The average snow temperature (K) (by layer mass) (NWM 3.0 output variable)
- `ISNOW` -- The number of snow layers (unitless) (NWM 3.0 output variable)
- `QRAIN` -- The rainfall rate on the ground (mm/s) (NWM 3.0 output variable)
- `FSNO` -- The snow-cover fraction on the ground (unitless fraction) (NWM 3.0 output variable)
- `SNOWH` -- The snow depth (m) (NWM 3.0 output variable)
- `SNLIQ` -- The snow layer liquid water (mm) (NWM 3.0 output variable)
- `QSNOW` -- The snowfall rate on the ground (mm/s) (NWM 3.0 output variable)
- `ECAN` -- Evaporation of intercepted water (mm) (NWM 3.0 output variable)
- `GH` -- The heat flux into the soil (W/m-2) (NWM 3.0 output variable)
- `TRAD` -- The surface radiative temperature (K) (NWM 3.0 output variable)
- `FSA` -- The total absorbed SW radiation (W/m-2) (NWM 3.0 output variable)
- `CMC` -- The total canopy water (liquid + ice) (mm) (NWM 3.0 output variable)
- `LH` -- The total latent heat to the atmosphere (W/m-2) (NWM 3.0 output variable)
- `FIRA` -- The total net LW radiation to atmosphere (W/m-2) (NWM 3.0 output variable)
- `FSH` -- The total sensible heat to the atmosphere (W/m-2) (NWM 3.0 output variable)

Listings for the above variables were added to the following `bmi/bmi_noahowp.f90` subroutines:

- `noahowp_var_grid`
  - All variables were assigned to the rank 2 grid, except for `SNLIQ`, for which a new instance of `GridType` (in `bmi/bmi_grid.f90`) was initialized with rank 3
- `noahowp_grid_type`
  - Added `GridType%id` = 2 (`GridType%rank` = 3) to uniform_rectilinear listing so that `noahowp_grid_type` works for `SNLIQ` and other 3D variables
  - All other variables should work as-is 
- `noahowp_var_type`
  - All variable added to listing for `real` except of `ISNOW`, which was added as `integer`
- `noahowp_var_units`
  - Some variables added and some variables have a units change
- `noahowp_var_itemsize`
- `noahowp_get_int`
  - (Only for `ISNOW`)
- `noahowp_get_float`
  - (All except `ISNOW`)

`ACSNOM` was added to the `watergrid_type` and `water_type` derived data types as well as the `WaterVarInTransfer` and `WaterVarOutTransfer` subroutines. Added calculation of `ACSNOM` in `WaterMain` subroutine in `src/WaterModule.f90`

`SNOWT_AVG` was added to the `energygrid_type` and `energy_type` derived data types as well as the `EnergyVarInTransfer` and `EnergyVarOutTransfer` subroutines. Added calculation of `SNOWT_AVG` in the `SnowWater` subroutine in `SnowWaterModule.f90`.

# Changes

This PR adds an additional instance of `GridType `to the `bmi_noahowp` derived type definition in `/bmi/bmi_noahowp.f90`. The additional instance corresponds to `GridType%rank = 3` and was needed because `SNLIQ` was added as a BMI-exposed variable and is 3D.

The PR modifies the BMI test program (i.e., `noahmp_driver_test` in `/test/noahowp_driver_test.f90`) to work the the newly added output variables. Modifications were necessary largely because the newly added `SNLIQ` variable is 3D and therefore needed to be handled differently than the other variables.

# Testing

I re-compiled and re-executed `/test/noahowp_driver_test.f90` and got [these results](https://github.com/NOAA-OWP/noah-owp-modular/files/12833176/noahowp_test_output.txt), which appear correct.

I also re-compiled and re-executed `/run/noah_owp_modular.exe` and got [these results](https://github.com/NOAA-OWP/noah-owp-modular/files/12833171/compare_outputnc.txt), which are correct given that cell (1,1) has the same input values as listed in `namelist.input` of the main branch of this repo.

# Notes

`SNLIQ`'s output_grid value should be 2 (i.e., rank 2). output_grid is currently set to 1 where `output_grid` is defined. I set `output_grid(14) = 2` where the value 14 corresponds to `SNLIQ`.

### CODE REVIEW

**Comment by GreyEvenson-NOAA** [comment]
@nmizukami, would you have time available to review this PR in the near future?

**Comment by GreyEvenson-NOAA** [comment]
@hellkite500: FYI I added another instance of `GridType` for rank 3 grids in this PR. Just calling your attention to this in case you'd like to check that I implemented correctly.

**Comment by nmizukami** [comment]
> @nmizukami, would you have time available to review this PR in the near future?

yes, I can do review!

**Comment by hellkite500** [comment]
> @hellkite500: FYI I added another instance of `GridType` for rank 3 grids in this PR. Just calling your attention to this in case you'd like to check that I implemented correctly.

Took a quick peek -- looks like the grid module is doing what I had hoped!  But I'll let you tell me whether or not adding that and using those mechanics was intuitive/simple enough or if it was the opposite!

File: test/noahowp_driver_test.f90 [hunk 1]
```diff
@@ -205,159 +225,191 @@ program noahmp_driver_test
...
-      ! Address real and integer variables differently. Do real variables now
-      if(var_type=='real') then
-
-        ! Allocate record keeping arrays
-        if(.NOT.allocated(var_value_get_real)) allocate(var_value_get_real(grid_size))
-        if(.NOT.allocated(var_value_set_real)) allocate(var_value_set_real(grid_size))
-        if(.NOT.allocated(grid_temp_real)) allocate(grid_temp_real(n_x,n_y))
-        if(size(var_value_get_real).ne.grid_size) then
-          deallocate(var_value_get_real)
+      print*,'    rank = ',grid_rank
```
Comment (nmizukami) [inline test/noahowp_driver_test.f90:233]: So here, there are mostly just print statement (so no formatted writing), but there is one write statement with format, and looking at test output (noahowp_test_output.txt). 

```
EVAPOTRANS
     rank =            2
     column count (n_x) =            2
     row count (n_y) =            3
     get_value function cpu time (s) =   1.90734863E-06
     from get_value (flattened) =
   0.00000000       0.00000000       0.00000000       0.00000000       8.58270573E-12   0.00000000    
     from get_value (in XY) =
       0.0        0.0  
       0.0        0.0  
       0.0        0.0  
```

there is small number 8.58270573E-12 in flatted output, but 0.0 in gridded output. 

Wonder if formatted output should be used for everything here?

**Review by nmizukami — COMMENTED** [review]
Hi @GreyEvenson-NOAA,    

Just looking through the PR.  I think overall it looks good to me and I added several picky comments!   Not super important but something to think about.

**Comment by GreyEvenson-NOAA** [comment]
Thank you, @nmizukami! Appreciate you taking the time. I'm out of town until Wednesday so will get on this on Thursday. Thanks again!

File: bmi/bmi_noahowp.f90 [hunk 2]
```diff
@@ -633,15 +658,21 @@ function noahowp_var_units(this, name, units) result (bmi_status)
...
-   case("PRCPNONC", "ETRAN")
+   case("PRCPNONC", "QRAIN", "QSEVA", "QSNOW")
       units = "mm/s"
       bmi_status = BMI_SUCCESS
-   case("SNEQV")
+   case("SNEQV", "ACSNOW", "EVAPOTRANS", "SNLIQ", "ECAN", "ETRAN", "CMC")
       units = "mm"
       bmi_status = BMI_SUCCESS
+   case("FSNO","ISNOW")
+      units = "1"
```
[the remark this answers is not in the document]
Comment (GreyEvenson-NOAA) [reply bmi/bmi_noahowp.f90:671]: I changed this to `units = "unitless"`

File: bmi/bmi_noahowp.f90 [hunk 3]
```diff
@@ -633,15 +658,21 @@ function noahowp_var_units(this, name, units) result (bmi_status)
...
+      bmi_status = BMI_SUCCESS
+   case("SNOWH")
+      units = "m"
+      bmi_status = BMI_SUCCESS
    case default
       units = "-"
```
[the remark this answers is not in the document]
Comment (GreyEvenson-NOAA) [reply bmi/bmi_noahowp.f90:677]: I'm actually not certain if "-" is meant to mean unitless as I'm thinking that the default case is intended to be used when `noahowp_var_units` is called for a variable name that isn't in `input_items` or `output_items` (?). So I was thinking we'd leave this as-is but make a change for `FSNO` and `ISNOW`. Let me know if you see potential problems in this approach.

File: bmi/bmi_noahowp.f90 [hunk 4]
```diff
@@ -837,13 +930,13 @@ function noahowp_get_float(this, name, dest) result (bmi_status)
       dest = reshape(watergrid%qinsur,[n_x*n_y])
       bmi_status = BMI_SUCCESS
    case("ETRAN")
-      dest = reshape(watergrid%etran,[n_x*n_y])
+      dest = reshape(watergrid%etran*domaingrid%DT,[n_x*n_y])
       bmi_status = BMI_SUCCESS
    case("QSEVA")
-      dest = reshape(watergrid%qseva,[n_x*n_y])
+      dest = reshape(watergrid%qseva*1000.,[n_x*n_y])
```
[the remark this answers is not in the document]
Comment (GreyEvenson-NOAA) [reply bmi/bmi_noahowp.f90:936]: I added these parameters and changed this listing to: `dest = reshape(watergrid%qseva*m2mm,[n_x*n_y])`

File: bmi/bmi_noahowp.f90 [hunk 5]
```diff
@@ -837,13 +930,13 @@ function noahowp_get_float(this, name, dest) result (bmi_status)
...
       bmi_status = BMI_SUCCESS
    case("EVAPOTRANS")
-      dest = reshape(watergrid%evapotrans,[n_x*n_y])
+      dest = reshape(watergrid%evapotrans*domaingrid%DT*0.001,[n_x*n_y])
```
[the remark this answers is not in the document]
Comment (GreyEvenson-NOAA) [reply bmi/bmi_noahowp.f90:939]: I changed this listing to `dest = reshape(watergrid%evapotrans*domaingrid%DT*mm2m,[n_x*n_y])`

File: src/WaterGridType.f90 [hunk 6]
```diff
@@ -50,6 +50,7 @@ module WaterGridType
   real,allocatable,dimension(:,:)                             :: SNEQVO      ! snow water eqv. of previous time step [mm]
   real,allocatable,dimension(:,:)                             :: BDSNO       ! bulk density of snowpack (kg/m3)
   real,allocatable,dimension(:,:)                             :: QSNBOT      ! melting water out of snow bottom [mm/s]
+  real,allocatable,dimension(:,:)                             :: ACSNOM      ! Accumulated meltwater from bottom snow layer [mm] (NWM 3.0)
   real,allocatable,dimension(:,:)                             :: PONDING
   real,allocatable,dimension(:,:)                             :: PONDING1
   real,allocatable,dimension(:,:)                             :: PONDING2
```
[the remark this answers is not in the document]
Comment (GreyEvenson-NOAA) [reply src/WaterGridType.f90:56]: I added comments for these variables such that these lines were revised to:

```
  real,allocatable,dimension(:,:)                             :: PONDING     ! surface ponding [mm] from snowmelt when thin snow has no layer
  real,allocatable,dimension(:,:)                             :: PONDING1    ! surface ponding [mm] from thin snow liquid during layer combination
  real,allocatable,dimension(:,:)                             :: PONDING2    ! surface ponding [mm] from thin snow liquid during transition from multilayer to no layer
```

These variable descriptions came from page 263 of He, C., Valayamkunnath, P., Barlage, M., Chen, F., Gochis, D., Cabell, R., … Ek, M. (2023). The Community Noah-MP Land Surface Modeling System Technical Description Version 5.0 (No. NCAR/TN-575+STR). doi:10.5065/ew8g-yr95

File: src/WaterModule.f90 [hunk 7]
```diff
@@ -108,6 +108,8 @@ SUBROUTINE WaterMain (domain, levels, options, parameters, forcing, energy, wate
     ! convert units (mm/s -> m/s)
     !PONDING: melting water from snow when there is no layer
     water%QINSUR = (water%PONDING+water%PONDING1+water%PONDING2)/domain%DT * 0.001
```
[the remark this answers is not in the document]
Comment (GreyEvenson-NOAA) [reply src/WaterModule.f90:110]: As there were multiple conversions (mm->m and m->mm) in this module, I revised multiple lines of src/WaterModule.f90 to use these parameters:

```
  real, parameter :: mm2m = 0.001  ! unit conversion mm to m
  real, parameter :: m2mm = 1000.  ! unit conversion m to mm
```

File: test/noahowp_driver_test.f90 (hunk 1 above)
Comment (GreyEvenson-NOAA) [reply test/noahowp_driver_test.f90]: I revised `/test/noahowp_driver_test.f90` `write` statement formatting (here and throughout) per your suggestion. Let me know if you think the formatting could be improved further.

Regarding `EVAPOTRANS` specifically, the revised program writes:

```
 EVAPOTRANS
 rank =            2
 column count (n_x) =            2
 row count (n_y) =            3
 get_value function cpu time (s) =   0.00000000    
 from get_value (flattened) =
        0.00000         0.00000         0.00000         0.00000         0.00000         0.00000  
 from get_value (in XY) =
        0.00000         0.00000  
        0.00000         0.00000  
        0.00000         0.00000  
 our replacement value (flattened) = 
        1.00000         2.00000         3.00000         4.00000         5.00000         6.00000  
 our replacement value (in XY) =
        1.00000         2.00000  
        3.00000         4.00000  
        5.00000         6.00000  
 set_value function cpu time (s) =   1.90734863E-06
 and the new value (flattened) = 
        1.00000         2.00000         3.00000         4.00000         5.00000         6.00000  
 and the new value (in XY) = 
        1.00000         2.00000  
        3.00000         4.00000  
        5.00000         6.00000  
```

**Comment by GreyEvenson-NOAA** [comment]
Regarding the conservation factor variables `m2mm` and `mm2m`:

The compiler did not allow me to declare `m2mm` and `mm2m` as `parameter` in both the `bminoahowp` and `WaterModule` modules because `bminoahowp` uses `WaterModule` (via its use of `RunModule`). 

The correct thing to do would seem to be to add `m2mm` and `mm2m` to `ConstantsModule` and use it in `bminoahowp` and `WaterModule`. However, I held off on that as I thought this would be better addressed by a separate PR for issue #5.

Hence, rather than adding `m2mm` and `mm2m` as parameters, I declared them as local variables within the scope of the `noahowp_get_float`, `noahowp_set_float`, and `WaterMain` subroutines.

Let me know if you think we should approach this a different way or if you see a better solution short of implementing `ConstantsModule`.

**Comment by GreyEvenson-NOAA** [comment]
@nmizukami, I believe we're ready for you to have another look. I appreciate your time and input!

File: bmi/bmi_noahowp.f90 [hunk 8]
```diff
@@ -1194,7 +1197,7 @@ function noahowp_set_float(this, name, src) result (bmi_status)
       watergrid%etran = reshape(src/domaingrid%DT,[n_x,n_y])
       bmi_status = BMI_SUCCESS
    case("QSEVA")
-      watergrid%qseva = reshape(src,[n_x,n_y])
+      watergrid%qseva = reshape(src*mm2m,[n_x,n_y])
```
Comment (nmizukami) [inline bmi/bmi_noahowp.f90:1200]: good to catch a small bug!?

**Review by nmizukami — APPROVED** [review]
Thank you @GreyEvenson-NOAA. It looks good to me. 

For the constant variables, yes, another PR to deal sounds good.  

For the format, test result looks better now, but you might also want to look into G format (I think it is more general), but for this PR, i think it is ok as is (if you find G-format better and it easy to implement it, would go ahead and change now)

**Comment by GreyEvenson-NOAA** [comment]
> Thank you @GreyEvenson-NOAA. It looks good to me.
> 
> For the constant variables, yes, another PR to deal sounds good.
> 
> For the format, test result looks better now, but you might also want to look into G format (I think it is more general), but for this PR, i think it is ok as is (if you find G-format better and it easy to implement it, would go ahead and change now)

Thanks for the tip! I implemented the G format codes per your suggestion. Nice and simple.

**Pull request merged**